### PR TITLE
Issue 18411 drag and drop images in containers in edit mode

### PR DESCRIPTION
**Pull request by alfredo-dotcms**
### Proposed Changes
* Drag and drop images into containers in edit mode

### Checklist
- [ ] We need to inject JavaScript code in the iframe to allow to do this (apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts)
- [ ] We upload the image to the temp resource
- [ ] We create a dotasset with the image
- [ ] We add the dotasset contentlet to the container

### Additional Info
** any additional useful context or info **

### Screenshots
Original             |  Updated
:-------------------------:|:-------------------------:
** original screenshot **  |  ** updated screenshot **

### CODE REVIEW

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.ts [hunk 1]
```diff
@@ -198,23 +201,34 @@ export class DotEditContentHtmlService {
...
-            // eslint-disable-next-line max-len
             `[data-dot-object="container"][data-dot-identifier="${this.currentContainer.identifier}"][data-dot-uuid="${this.currentContainer.uuid}"]`
         );
 
         if (this.isContentExistInContainer(contentlet, containerEl)) {
             this.showContentAlreadyAddedError();
         } else {
-            const contentletPlaceholder = this.getContentletPlaceholder();
-            containerEl.appendChild(contentletPlaceholder);
+            let contentletPlaceholder;
```
Comment (hmoreras) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.ts:224]: `contentletPlaceholder` is defined here and in line 285

**Review by hmoreras — CHANGES_REQUESTED** [review]

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.spec.ts [hunk 2]
```diff
@@ -44,7 +45,7 @@ const mouseoverEvent = new MouseEvent('mouseover', {
     cancelable: true
 });
 
-xdescribe('DotEditContentHtmlService', () => {
+describe('DotEditContentHtmlService', () => {
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.spec.ts:48]: This test was failing intermittently :(

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 3]
```diff
@@ -143,6 +143,240 @@ export const EDIT_PAGE_JS = `
             return this.down && drake.dragging;
         }
     });
+
+    // D&D DotAsset - Start
+
+    function dotAssetCreate(options) {
+        var promises = [];
+        var filesCreated = 1;
+        options.files.map((file) => {
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:152]: Are we supporting multiple files here?

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 4]
```diff
@@ -143,6 +143,240 @@ export const EDIT_PAGE_JS = `
...
+                    hostFolder: options.folder,
+                    indexPolicy: 'WAIT_FOR'
+                }
+            };
+
+            promises.push(
+                fetch(options.url, {
+                    method: 'PUT',
+                    headers: {
+                        Origin: window.location.hostname,
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:166]: required?

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 5]
```diff
@@ -143,6 +143,240 @@ export const EDIT_PAGE_JS = `
...
+            if (errors.length) {
+                throw errors;
+            } else {
+                return data;
+            }
+        });
+    }
+
+    function uploadBinaryFile(file, maxSize) {
+        var path = '/api/v1/temp';
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:202]:
```suggestion
        const path = '/api/v1/temp';
```

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 6]
```diff
@@ -143,6 +143,240 @@ export const EDIT_PAGE_JS = `
...
+        path += maxSize ? '?maxFileLength=' + maxSize : '';
+        var formData = new FormData();
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:204]:
```suggestion
        const formData = new FormData();
```

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 7]
```diff
@@ -143,6 +143,240 @@ export const EDIT_PAGE_JS = `
...
+                    message: (await response.json()).message,
+                    status: response.status
+                };
+                throw error;
+            }
+        });
+    }
+
+    function uploadFile(file, maxSize) {
+        if (typeof file !== 'string') {
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:226]: Can't we do `=== to something`?

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 8]
```diff
@@ -143,6 +143,240 @@ export const EDIT_PAGE_JS = `
...
+
+    function setLoadingIndicator() {
+        var currentContentlet = document.getElementById('contentletPlaceholder');
+        currentContentlet.classList.remove('gu-transit');
+        currentContentlet.innerHTML = '<div class="loader__overlay"><div class="loader"></div></div>';
+    }
+
+    function isCursorOnUpperSide(cursor, contentletBoundingRect) {
+        return cursor.y - contentletBoundingRect.top  <  (contentletBoundingRect.bottom - contentletBoundingRect.top)/2
+    }
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:239]:
```suggestion
    function isCursorOnUpperSide(cursor, { top, bottom }) {
        return cursor.y - top  <  (bottom - top)/2
    }
```

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 9]
```diff
@@ -143,6 +143,240 @@ export const EDIT_PAGE_JS = `
...
+
+    function isContentletPlaceholderInDOM() {
+        return document.getElementById('contentletPlaceholder');
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:242]:
```suggestion
        return !!document.getElementById('contentletPlaceholder');
```

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 10]
```diff
@@ -143,6 +143,240 @@ export const EDIT_PAGE_JS = `
...
+        // Container does NOT allow img
+        if (!container.dataset.dotAcceptTypes.toLocaleLowerCase().match(/dotasset/g)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    var contentletPlaceholder;
+    var currentContainer;
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:286]:
```suggestion
    let contentletPlaceholder;
    let currentContainer;
```

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 11]
```diff
@@ -143,6 +143,240 @@ export const EDIT_PAGE_JS = `
...
+            contentletPlaceholder = document.createElement('div');
+            contentletPlaceholder.id = 'contentletPlaceholder';
+            contentletPlaceholder.setAttribute('data-dot-object', 'contentlet')
+            contentletPlaceholder.classList.add('gu-transit')
+            
+            if (isContainerAndContentletValid(container, contentlet) && isCursorOnUpperSide(event, contentlet.getBoundingClientRect())) {
+                insertBeforeElement(contentletPlaceholder, contentlet);
+            } else if (isContainerAndContentletValid(container, contentlet)) {
+                insertAfterElement(contentletPlaceholder, contentlet);
+            }
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:325]:
```suggestion
            if (isContainerAndContentletValid(container, contentlet) {
                if (isCursorOnUpperSide(event, contentlet.getBoundingClientRect())) {
                    insertBeforeElement(contentletPlaceholder, contentlet);
                } else {
                    insertAfterElement(contentletPlaceholder, contentlet);                    
                }
            }
```

**Review by fmontes — CHANGES_REQUESTED** [review]

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.spec.ts (hunk 2 above)
Comment (alfredo-dotcms) [reply apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.spec.ts:48]: mmm, this time it worked!
image

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.ts (hunk 1 above)
Comment (alfredo-dotcms) [reply apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.ts:224]: I don't see a valid reason why

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts (hunk 4 above)
Comment (alfredo-dotcms) [reply apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:166]: no

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts (hunk 5 above)
Comment (alfredo-dotcms) [reply apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:202]: needs to be `let`

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.spec.ts (hunk 2 above)
Comment (alfredo-dotcms) [reply apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.spec.ts:48]: 2nd time after feedback!
image

**Review by hmoreras — APPROVED** [review]

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 12]
```diff
@@ -143,6 +143,230 @@ export const EDIT_PAGE_JS = `
...
+                asset: options.file.id,
+                hostFolder: options.folder,
+                indexPolicy: 'WAIT_FOR'
+            }
+        };
+
+        return fetch(options.url, {
+            method: 'PUT',
+            headers: {
+                Origin: window.location.hostname,
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:162]: We need this?

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 13]
```diff
@@ -143,6 +143,230 @@ export const EDIT_PAGE_JS = `
...
+                return (await response.json()).tempFiles[0];
+            } else {
+                const error = {
+                    message: (await response.json()).message,
+                    status: response.status
+                };
+                throw error;
+            }
+        });
+    }
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:211]: What do we show in the UI if there is an error here?

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 14]
```diff
@@ -143,6 +143,230 @@ export const EDIT_PAGE_JS = `
...
+
+    function insertBeforeElement(newElem, element) {
+        element.parentNode.insertBefore(newElem, element);
+    }
+
+    function insertAfterElement(newElem, element) {
+        try{
+            element.parentNode.insertBefore(newElem, element.nextSibling);
+        }catch(error){}
+    }
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:245]: Is error we die silently here, are we providing any UI error to the user?

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 15]
```diff
@@ -143,6 +143,230 @@ export const EDIT_PAGE_JS = `
...
+        event.stopPropagation();
+
+        const container = event.target.closest('[data-dot-object="container"]');
+        currentContainer = container;
+
+        if (container && !checkIfContainerAllowsDotAsset(event)) {
+            container.classList.add('no');
+        }
+
+    }, false);
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:288]: Don't we need to remove these listeners when we pass from edit to preview?

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts [hunk 16]
```diff
@@ -143,6 +143,230 @@ export const EDIT_PAGE_JS = `
...
+                }).then((response) => {
+                    window.contentletEvents.next({
+                        name: 'add-uploaded-dotAsset',
+                        data: {
+                            contentlet: response,
+                            placeholderId: 'contentletPlaceholder'
+                        }
+                    });
+                })
+            })
```
Comment (fmontes) [inline apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/libraries/iframe-edit-mode.js.ts:358]: If there is an error here, what do we show to the user?

**Review by fmontes — CHANGES_REQUESTED** [review]

File: apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.spec.ts (hunk 2 above)
Comment (alfredo-dotcms) [reply apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.spec.ts:49]: 3rd time after more feedback
image

**Review by fmontes — APPROVED** [review]

**Review by hmoreras — APPROVED** [review]

**Pull request merged**